File: 06-week-5--big-o-continued/00-days-1-to-2--implement-a-stack-class/javascript/stack.js
class Stack {
  constructor() {
    this.stack = [];
    // this is an arbitrary value to make testing easier
    // 1,024 would be way too high to test!
    this.limit = 10;
  }

  // add item to top of stack if not full
  // if full throw error
  push(item) {
    if (!this.isFull()) {
      return this.stack.push(item)
    } else {
      throw new Error('The Stack Is Full!')
    }
  }

  // remove item from top of stack and return it
  pop() {
    return this.stack.pop()
  }

  // return item at top of stack without removing it
  peek() {
    return !!this.stack.length > 0 ? this.stack[this.stack.length - 1] : undefined
  }

  // return true if stack is empty, otherwise false
  isEmpty() {
    return this.size() === 0
  }

  // return true if stack is full, otherwise false
  isFull() {
    return this.stack.length >= this.limit
  }

  // return number of items in stack
  size() {
    return this.stack.length
  }

  // return -1 if item not in stack, otherwise integer representing 
  // how far it is from the top
  search(target) {
    return !!this.stack.find(element => element === target) ? (this.stack.length - 1) - this.stack.findIndex(element => element === target) : -1
  }

  // print contents of stack: do not return the stack itself!
  print() {
    console.log(this.stack)
  }
}

if (require.main === module) {
  // add your own tests in here
}

module.exports = Stack;

File: 06-week-5--big-o-continued/00-days-1-to-2--implement-a-stack-class/ruby/stack.rb
class Stack
  attr_reader :limit

  def initialize
    @stack = []
    # this is an arbitrary value to make testing easier
    # 1,024 would be way too high to test!
    @limit = 10
  end

  # add item to top of stack if not full
  # if full, throw error
  def push(item)
    if !isFull?
      @stack.push(item)
    else
      raise 'Stack is full'
    end
  end

  # remove item from top of stack and return it
  def pop
    @stack.slice!(size - 1)
  end

  # return item at top of stack without removing it
  def peek
    self.stack.last
  end

  # return true if stack is empty, otherwise false
  def isEmpty?
    self.size === 0
  end

  # return true if stack is full, otherwise false
  def isFull?
    size === @limit
  end

  # return number of items in stack
  def size
    @stack.length
  end

  # return -1 if item not in stack, otherwise integer representing 
  # how far it is from the top
  def search(target)
    if @stack.include?(target)
      (size - 1) - @stack.index(target)
    else
      -1
    end
  end

  # print contents of stack: do not return the stack itself!
  def print
    puts @stack
  end
end

if __FILE__ == $PROGRAM_NAME
  # Don't forget to add your tests!
end

